### PR TITLE
[FIRRTL] Handle RegReset as index to SubAccess

### PR DESCRIPTION
**Pull request by prithayan**
`InferResets`  was not handling the case when the reset reg is an `index` into the `SubAccess` Op and not the `input`.
Fixes https://github.com/llvm/circt/issues/2132

### CODE REVIEW

File: lib/Dialect/FIRRTL/Transforms/InferResets.cpp [hunk 1]
```diff
@@ -206,13 +206,17 @@ static bool insertResetMux(ImplicitLocOpBuilder &builder, Value target,
...
-        .Case<SubaccessOp>([&](auto op) {
-          auto resetSubValue =
-              builder.create<SubaccessOp>(resetValue, op.index());
+        .Case<SubaccessOp>([&](SubaccessOp op) {
+          Value resetSubValue;
+          if (op.index() == target)
+            resetSubValue = builder.create<SubaccessOp>(op.input(), resetValue);
+          else
+            resetSubValue = builder.create<SubaccessOp>(resetValue, op.index());
+
```
Comment (fabianschuiki) [inline lib/Dialect/FIRRTL/Transforms/InferResets.cpp:215]:
```suggestion
        .Case<SubaccessOp>([&](auto op) {
          if (op.input() != target)
            return;
          auto resetSubValue =
              builder.create<SubaccessOp>(resetValue, op.index());
```

File: lib/Dialect/FIRRTL/Transforms/InferResets.cpp [hunk 2]
```diff
@@ -206,13 +206,17 @@ static bool insertResetMux(ImplicitLocOpBuilder &builder, Value target,
...
           if (insertResetMux(builder, op, reset, resetSubValue))
             resetValueUsed = true;
           else
-            resetSubValue.erase();
+            resetSubValue.getDefiningOp()->erase();
```
Comment (fabianschuiki) [inline lib/Dialect/FIRRTL/Transforms/InferResets.cpp:219]:
```suggestion
            resetSubValue.erase();
```

**Review by fabianschuiki — CHANGES_REQUESTED** [review]
Thanks for tackling this @prithayan! I think all that's needed is a
```
if (op.input() != target)
  return;
```
as the first statement in the case for `SubaccessOp`. We explicitly don't want to insert multiplexers in front of the index value of a subaccess, because the code basically tries to find subaccesses into registers that need to get a reset, which means descending down the LHS/input side of the subaccesses.

Love the test case! :+1:

**Review by fabianschuiki — APPROVED** [review]
Looks great! :tada:

**Comment by youngar** [comment]
Do we need the same check for `SubfieldOp` and `SubindexOp`?

**Comment by prithayan** [comment]
> 
> 
> Do we need the same check for `SubfieldOp` and `SubindexOp`?

I thought same, but the index is an Integer attribute for the `SubfieldOp` and `SubindexOp`, so this case cannot occur.

**Comment by youngar** [comment]
Ahh yeah that makes sense! thanks!

**Pull request merged**